### DISCUSSION TITLE
Deactivation of control in angle deviation

### DISCUSSION THREAD

**Discussion by killinen** [dzid26/StepperServoCAN#18] [Ideas]
If used on a steering column, it is crucial that your StepperServoCAN assembly remains intact, with none of the parts coming loose.

This safety enhancement would be highly beneficial and could be easily implemented if the car already has a steering angle sensor. By monitoring the maximum deviation between the car's steering angle sensor and the StepperServoCAN angle, we can ensure safe and precise control. If the deviation exceeds a predefined threshold, the StepperServoCAN would no longer accept control messages.

To achieve this, we could introduce a fifth STEER_MODE, such as the "deviation tracking mode." Users can choose to utilize this mode optionally, and the system would not enforce its usage. In such a mode, the maximum deviation event could be reported in the CONTROL_STATUS message.

However, there are a few challenges to consider. One of them is efficiently incorporating the required CAN encodings of the steering angle into the StepperServoCAN firmware. Additionally, the maximum acceptable deviation might need to be user-defined since different assemblies may have varying tolerances. Implementing these features could increase the firmware's size, so we could address this issue on the OP side.

**Comment by dzid26** [comment]
> Implementing these features could increase the firmware's size so that we could address this issue on the OP side.

Right, this mostly belongs to the OP side (or Panda), not so much because of size but because it's application specific and adds complexity to interfaces, etc.  If deviation is detected, OP shall request SOFT_OFF. 

(This would only miss unlikely hardware failures when the motor is not actively used. )

To make sure the angle from the motor and the car has the same offset, a simple zeroing could be part of calibration (or a dedicated function could called by some means of a button, CAN or debugger...).  (I have some simple code WIP for zeroing after calibration and storing it in eeprom that I can publish soon).  

Then the savvy user must still ensure the zero doesn't change by full rotation. This can by achieved means of either always powering down with steering close to the center or never powering down (new StepperServoCAN will be more energy efficient). - Otherwise, the deviation safety would trigger in OP.

**Reply by killinen** [reply]
Yes I agree that it probably is better to do on OP side. The benefit doing it on StepperServoCAN side could be that it would offload some of development from incorporating the StepperServoCAN into OP. Not even sure of that, could be easily done with cherry-picking etc. 

> To make sure the angle from the motor and the car has the same offset, a simple zeroing could be part of calibration

I think I have little different way achieving this. At carstate I'll check `angle_offset` of the StepperServoCAN angle `steeringAngleDegSSC` and at every startup (leveredged the ZSS code), and then track the min/max difference `steeringAngleDegDivergence`. If big enough divergence is detected OP will prompt a warning.

If I recall correct the code in the link did not work correctly but it gives idea what I'm doing (working code on device that I don't have ATM on hand).
https://github.com/killinen/openpilot/commit/57b1db356e423341504bd5dcef3188d470b9e69f

**Reply by dzid26** [reply]
> I think I have little different way achieving this. 

You right. It's easier to simply set the offset in the OP on powerup.  I am mixing a bit the idea of using a motor sensor as a primary steering sensor for people who don't have one already.  Anyway, both solutions don't exclude each other. 

> Not even sure of that, could be easily done with cherry-picking etc.

Yeah, one of the issues is that there are different cars in `selfdrive/car`.   Maybe [git patch](https://stackoverflow.com/questions/47264829/is-it-possible-to-create-a-git-patch-from-one-directory-and-apply-that-patch-to) willbe useful, albeit it is only 2-way merge.  
I will upgrading to the latest OP this year, so I will be thinking about this problem.

**Reply by killinen** [reply]
> I am mixing a bit the idea of using a motor sensor as a primary steering sensor for people who don't have one already.

Yeah, this will be value adding feature. Great see and hear that you are still developing the HW/SW. Good luck for the upgrade, I'm doing the same (at least what will work on Freon).
